### DISCUSSION TITLE
Is there a cleanup function for alpinejs components when they are removed from the dom

### DISCUSSION THREAD

**Discussion by EthanBarlo** [livewire/livewire#9062] [1. Help] [answered]
In some of my components I create event listeners. 
One good example is I have a resizable grid component.
Which calculates the optimal number of columns based on the total_items, avaliable_space and min_item_width.
SO I have an event listener for 

```js
const handleResize = () => this.handleResize();
window.addEventListener('resize', handleResize)
```

We've started using `wire:navigate` around the application and noticed that these events are sticking around which makes sense since we're not reloading the page anymore.

Is there something we can hook into within Alpine or Livewire, when a component is removed from the page.
So that we can run a cleanup?

I suppose I could add another listener for `livewire:navigating` or `livewire:navigated`
```js
const handleResize = () => this.handleResize();
window.addEventListener('resize', handleResize);

// Use livewire:navigating to run the cleanup
document.addEventListener('livewire:navigating', () => {
    // Check if the $el exists
    if(this.$el !== null){
        window.removeEventListener('resize', handleResize())
   }
},{ once: true })
```
The issue with this approach is it would only work on navigation.
Sometimes components could be added or removed with livewire.

So I wondered whether there was something we could hook into to detect when an alpine component is about to be removed from the dom.

**Comment by joshhanley** [comment, marked as answer]
@EthanBarlo yeah a `wire:navigating` listener is best for cleaning up any listeners you have set up when navigating.

But I believe if you set up your listeners in a `component.init` hook https://livewire.laravel.com/docs/javascript#component-initialization then the second parameter `cleanup` is a function where you can pass a function into, and Livewire will run that when the component is being cleaned up, so inside that function you could clean up your listeners.

Hope this helps!

**Reply by EthanBarlo** [reply]
@joshhanley Amazing this is exactly what I was looking for.

**Reply by hasan-aftab** [reply]
Josh brother u are a life saver <3
